h/auth/policy.py

```diff
@@ -41,34 +43,104 @@ def forget(self, request):
 
 
 @interface.implementer(interfaces.IAuthenticationPolicy)
-class AuthClientPolicy(BasicAuthAuthenticationPolicy):
+class AuthClientPolicy(object):
 
     """
-    An authentication policy for registered auth_clients.
+    An authentication policy for registered auth_clients
+
+    Authentication for a request to API routes with HTTP Basic Authentication
+    credentials that represent a registered AuthClient with client_credentials
+    in the db.
+
+    Authentication can be of two types:
+
+    * The client itself. Some endpoints allow an authenticated auth_client to
+      take action on users within its authority, such as creating a user or
+      adding a user to a group. In this case, assuming credentials are valid,
+      the request will be authenticated, but no ``authenticated_userid`` (and
+      thus no request.user) will be set
+    * A user within the client's associated authority. If an HTTP
+      `X-Forwarded-User` header is present, its value will be treated as a
+      ``userid`` and, if the client credentials are valid _and_ the userid
+      represents an extant user within the client's authority, the request
+      will be authenticated as that user. In this case, ``authenticated_userid``
+      will be set and there will ultimately be a request.user available.
+
+    Note: To differentiate between request with a Token-authenticated user and
+    a request with an auth_client forwarded user, the latter has an additional
+    principal, ``client:{client_id}@{authority}`` to mark it as being authenticated
+    on behalf of an auth_client
     """
 
-    def __init__(self, check, realm='Realm', debug=False):
-        self.check = check
-        self.realm = realm
-        self.debug = debug
+    def __init__(self, check=None):
+        if check is None:
+            check = AuthClientPolicy.check
```

Comment:
TIL: you can't default a constructor param for to a class method on the class for which you're creating an instance :smile: